runtime/src/accounts_db.rs

```diff
@@ -3660,79 +3656,205 @@ impl AccountsDB {
+    ) -> std::cmp::Ordering {
+        match a.pubkey.cmp(&b.pubkey) {
+            std::cmp::Ordering::Equal => match a.slot.cmp(&b.slot).reverse() {
+                std::cmp::Ordering::Equal => a.version.cmp(&b.version).reverse(),
+                other => other,
+            },
+            other => other,
+        }
+    }
+
+    fn sort_hash_intermediate(
+        mut data_by_pubkey: Vec<CalculateHashIntermediate>,
+    ) -> (Vec<CalculateHashIntermediate>, Measure) {
+        // sort each PUBKEY_DIVISION vec
+        let mut sort_time = Measure::start("sort");
+        data_by_pubkey.par_sort_unstable_by(Self::compare_two_hash_entries);
+        sort_time.stop();
+        (data_by_pubkey, sort_time)
+    }
+
+    fn de_dup_and_eliminate_zeros(
+        sorted_data_by_pubkey: Vec<CalculateHashIntermediate>,
+    ) -> (Vec<Vec<Hash>>, Measure, u64) {
+        // 1. eliminate zero lamport accounts
+        // 2. pick the highest slot or (slot = and highest version) of each pubkey
+        // 3. produce this output:
+        // vec: PUBKEY_BINS_FOR_CALCULATING_HASHES in pubkey order
+        //   vec: sorted sections from parallelism, in pubkey order
+        //     vec: individual hashes in pubkey order
+        let mut zeros = Measure::start("eliminate zeros");
+        const CHUNKS: usize = 10;
+        let (hashes, sum) = Self::de_dup_accounts_in_parallel(&sorted_data_by_pubkey, CHUNKS);
+        zeros.stop();
+        (hashes, zeros, sum)
+    }
+
+    // 1. eliminate zero lamport accounts
+    // 2. pick the highest slot or (slot = and highest version) of each pubkey
+    // 3. produce this output:
+    //   vec: sorted sections from parallelism, in pubkey order
+    //     vec: individual hashes in pubkey order
+    fn de_dup_accounts_in_parallel(
+        pubkey_division: &[CalculateHashIntermediate],
+        chunk_count: usize,
+    ) -> (Vec<Vec<Hash>>, u64) {
+        let len = pubkey_division.len();
+        let max = if len > chunk_count { chunk_count } else { 1 };
+        let chunk_size = len / max;
+        let overall_sum = Mutex::new(0u64);
+        let hashes: Vec<Vec<Hash>> = (0..max)
+            .into_par_iter()
+            .map(|chunk_index| {
+                let mut start_index = chunk_index * chunk_size;
+                let mut end_index = start_index + chunk_size;
+                if chunk_index == max - 1 {
+                    end_index = len;
+                }
+
+                let first_slice = chunk_index == 0;
+                if !first_slice {
+                    start_index -= 1;
+                }
+
+                let (result, sum) = Self::de_dup_accounts_from_stores(
+                    chunk_index == 0,
+                    &pubkey_division[start_index..end_index],
+                );
+                let mut overall = overall_sum.lock().unwrap();
+                *overall = Self::checked_cast_for_capitalization(sum + *overall as u128);
+
+                result
```

Comment:
Hmm this won't guarantee the most equitable distribution of work if the divisibility is an issue right? For instance, 19 work items, split among 10 workers here, would yield 9 works with 1 item, and 1 worker with 10 items here.  

How about something like this: https://github.com/carllin/solana/commit/9a8eb0482c2b951d34714193e17c7b8fed76eec8

Reply:
My thinking was: some divisibility is better than none. Completely equitable is great. Always correct is necessary. You are right about len = 19. It isn't good. My assumption was data sizes that are already large (15M). But the last worker's share arrives at an equitable amount quickly enough. len=10k gives you 10.08% for the last worker. For 1M, it is 10.0004%. I can certainly tweak the algorithm for distribution to use float division or other techniques. I am not confident this algorithm will remain implemented as it is for even 6 months. Already there is the 80M and 1B talk ;-) I imagine this to be a sufficiently simple, and empirically faster implementation as we learn more and find the next hotspots and come up with even better implementations.

Reply:
I propose that we treat the optimization of the size per worker selection as a follow on discussion to this PR. Is that reasonable?

Reply:
It's all good, the drift is at most `chunk_count` anyways, which is currently small. Was more for future-proofing in case I wanted such a utility in the future 😃